### ISSUE TITLE
Screenshot different on M1 vs x64 devices

### ISSUE THREAD

**Issue by matthew-shin-hootsuite** [ndtp/android-testify#101]
**Describe the bug**

- Running the same screenshot test on a Macbook Pro with M1 chip vs. Intel x64 results in failing tests. The diff between these two recordings look identical content wise so it seems like it's some sizing or some other thing that is off. 
- In order to get around this issue I tried to use the `setUseSoftwareRenderer(true)` as suggested in the doc to ignore the architecture differences but it didn't seem to help
- Also tried to reduce the exactness of each failing test to see if I could lower the tolerance to get it to pass but this turned out to be a bit of a flaky test as each time I run it different sets seem to fail, some errors such:

```androidx.test.espresso.NoActivityResumedException: No activities found. Did you forget to launch the activity by calling getActivity() or startActivitySync or similar?
        at dalvik.system.VMStack.getThreadStackTrace(Native Method)
        at java.lang.Thread.getStackTrace(Thread.java:1720)
        at androidx.test.espresso.base.DefaultFailureHandler.getUserFriendlyError(DefaultFailureHandler.java:12)
        at androidx.test.espresso.base.DefaultFailureHandler.handle(DefaultFailureHandler.java:7)
        at androidx.test.espresso.ViewInteraction.waitForAndHandleInteractionResults(ViewInteraction.java:8)
        at androidx.test.espresso.ViewInteraction.desugaredPerform(ViewInteraction.java:11)
        at androidx.test.espresso.ViewInteraction.perform(ViewInteraction.java:4)
        at androidx.test.espresso.Espresso.closeSoftKeyboard(Espresso.java:1)
        at dev.testify.ScreenshotRule.assertSame(ScreenshotRule.kt:493)
```

This issue relates to:
- [X] The Kotlin library
- [ ] The Gradle plugin
- [ ] The IntelliJ Platform plugin
- [ ] The sample code
- [ ] The documentation

**To Reproduce**
1. Use an M1 device to record a screenshot.
2. Confirm it passes on its own device
3. Re-run the same passing test on a non-M1 device
4. See a failing test when expecting a pass

**Expected behavior**
Expect the screenshots to pass regardless of the architecture of the processor of a device

**Desktop (please complete the following information):**
 - OS: macOS
 - Version 13.0.1

**Target Android Device (please complete the following information):**
 - Device: Pixel 3a
-  Physical or Virtual: Android Virtual Device (Emulator)
 - API Level: 29
- Testify Key: 29-1080x2220@440dp-en_US

**Comment by DanielJette** [comment]
@matthew-shin-hootsuite What is the main UI toolkit you are using in your app? Android-based Views or Compose?

Are you using any Espresso actions with your tests, or just the built-in integration with Testify. I see some Espresso errors in the log you posted and wondering if that's solely coming from inside Testify, or if you are also running some Espresso code in your tests.

Finally, which [Capture Method](https://ndtp.github.io/android-testify/docs/recipes/capture-method) are you using with your tests?

**Comment by matthew-shin-hootsuite** [comment]
> Are you using any Espresso actions with your tests, or just the built-in integration with Testify.

No espresso, just the one built in from Testify. 

> @matthew-shin-hootsuite What is the main UI toolkit you are using in your app? Android Views or Compose?


We have both a combination of Compose and Android Views and both tests have failing test. 

> Finally, which [Capture Method](https://ndtp.github.io/android-testify/docs/recipes/capture-method) are you using with your tests?

For the capture method we use the default `DrawingCache` on Android Views and `PixelCopyCapture` for [ComposeTests](https://github.com/ndtp/android-testify/blob/main/Ext/Compose/src/main/java/dev/testify/ComposableScreenshotRule.kt#L60).


For your reference, I actually also downloaded the Testify repo and ran the `screenshotTest` suite and also confirmed it fails. Here is the stack trace for one of the failed tests:

```27) default(dev.testify.sample.clients.details.ClientDetailsViewScreenshotTest)
dev.testify.internal.exception.ScreenshotIsDifferentException: 

*  The captured screenshot is different from the baseline screenshot.
*  Run `./gradlew Sample::screenshotPull` to view the differences.
*  Run `./gradlew Sample::screenshotTest -PtestClass=dev.testify.sample.clients.details.ClientDetailsViewScreenshotTest#default` to run this test again.


        at dev.testify.ScreenshotRule.assertSame(ScreenshotRule.kt:500)
        at dev.testify.sample.clients.details.ClientDetailsViewScreenshotTest.default(ClientDetailsViewScreenshotTest.kt:63)
        at java.lang.reflect.Method.invoke(Native Method)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at dev.testify.ScreenshotRule$ScreenshotStatement.evaluate(ScreenshotRule.kt:568)
        at androidx.test.rule.ActivityTestRule$ActivityStatement.evaluate(ActivityTestRule.java:549)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at androidx.test.ext.junit.runners.AndroidJUnit4.run(AndroidJUnit4.java:162)
        at org.junit.runners.Suite.runChild(Suite.java:128)
        at org.junit.runners.Suite.runChild(Suite.java:27)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
        at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
        at androidx.test.internal.runner.TestExecutor.execute(TestExecutor.java:56)
        at androidx.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:444)
        at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:2189)
```

As well, here is the `screenshotPull` from the same failing test. When referencing manually to the baseline, it looks the same.

![ScreenshotRuleExampleTests_default](https://user-images.githubusercontent.com/86132583/205753809-4feae6f9-7497-41b6-ba17-7517eef091bc.png)

**Comment by DanielJette** [comment]
Thanks for the additional information @matthew-shin-hootsuite 

I'll investigate this a bit further and get back to you as soon as I can

**Comment by DanielJette** [comment]
@matthew-shin-hootsuite I haven't had a chance to run the Sample on M1 & intel hardware yet, but as a quick check, can you try using a [matching tolerance](https://ndtp.github.io/android-testify/docs/recipes/tolerance) of `0.9f` for all of your screenshot tests?

Despite what the documentation says, the tolerance code has been multithreaded recently and so it's not that much slower than a direct match. I use this on all of my projects and it has successfully allowed us to run M1 tests locally and Intel x86 on our CI environment.


```kotlin
    @Before
    fun setUp() {
        rule.setExactness(0.9f)
    }
```

Or, if you are using `ComposableScreenshotRule`, you can specify the `exactness` in the constructor.

**Comment by matthew-shin-hootsuite** [comment]
> @matthew-shin-hootsuite I haven't had a chance to run the Sample on M1 & intel hardware yet, but as a quick check, can you try using a [matching tolerance](https://ndtp.github.io/android-testify/docs/recipes/tolerance) of `0.9f` for all of your screenshot tests?
> 
> Despite what the documentation says, the tolerance code has been multithreaded recently and so it's not that much slower than a direct match. I use this on all of my projects and it has successfully allowed us to run M1 tests locally and Intel x86 on our CI environment.
> 
> ```kotlin
>     @Before
>     fun setUp() {
>         rule.setExactness(0.9f)
>     }
> ```
> 
> Or, if you are using `ComposableScreenshotRule`, you can specify the `exactness` in the constructor.

Hey @DanielJette, I actually also have a customized set of rules one of which is to test in dark mode which uses the `ComposableScreenshotRule` and sets the exactness to `0.9f` already. Also while testing and trying to get it to work, I've went down to some of these tests with as low as `0.85f` to try and get them to pass the test but there are two main concerns here with this:

1. My understanding of exactness is pretty much the tolerance between the differences between the two screenshots and the lower it is, the higher the tolerance which won't be always good in other scenarios.
2. Even if you assume that the first issue is not a concern, I've run into problems where different sets of tests seem to be failing on different runs when I change these exactness, not sure why.

**Comment by DanielJette** [comment]
@matthew-shin-hootsuite Yeah, I get where you're coming from. For (1), the tolerance uses a [delta-E ](https://www.viewsonic.com/library/creative-work/what-is-delta-e-and-why-is-it-important-for-color-accuracy/) algorithm for comparing images, so it's specifically designed to ignore minor color variations but will fail on changes to positioning, size or layout of elements.

But, (2) is the real concern here. If we are dealing with problems due to differences in hardware, then this _should_ be addressable by using a moderate exactness of around 0.9. If you've gone up to 0.85 without success, that indicates that something else may be contributing to a more significant difference. I for sure what to investigate this more closely and hopefully can find some solid solutions for you.

Some things you can try in the meantime:
1. You can continuously lower the exactness until you get a pass. This would be only for additional knowledge about the problem. If you get down to, say 0.1f and it still fails, then that would be informative and help narrow down the problem.
2. You can [enable the `GenerateDiffs` Testify feature](https://github.com/ndtp/android-testify/blob/230607acc598afe7d54f9618d55fdecd0da83800/Sample/src/androidTest/java/dev/testify/sample/ScreenshotRuleExampleTests.kt#L234-L250). When enabled, this will output a `.diff.png` alongside the existing baseline images. These `diff` images are high-contrast image where each difference, regardless of how minor, is indicated as:
    - Black: No difference
    - Red: Pixel difference is greater than the allowable tolerance
    - Yellow: Pixel is different, but within the allowable tolerance
    - Gray: Pixel is excluded from the difference (not compared)
This will also give more information as to the cause of the failures and may help narrow down the possible solutions.

**Comment by matthew-shin-hootsuite** [comment]
@DanielJette thanks for that information, indeed what you mention is probably a concern. The output logs I get are so random that it's difficult to do the steps you asked properly to get real any useful data. Did you have any success in getting similar results with a M1 device? Since with the sample examples in the repo I was able to reproduce the issue, I think that's a pretty strong indication that it's not the tests that I'm writing but with incompatibility with M1. Thoughts?

**Comment by DanielJette** [comment]
@matthew-shin-hootsuite I was actually just getting ready to add some more information to this issue!

I ran some tests with:
 (1) Intel Core i7, Radeon RX 580, macOS 11.6
 (2) Apple M1 Max maxOS 12.2.1

I observed the same rendering differences as you did.
On closer analysis, it appears that the cause for the differences are all found in hardware accelerated parts of the graphics pipeline. This would be:
1. Shadows & elevation
2. Font smoothing & anti-aliasing
3. Image decompression and rendering
4. Alpha-blending

Most of these are present in the differences between an M1 and Intel CPU.

I was able to compensate for these errors by adding the following to the Sample application:

```kotlin
    @Before
    fun setUp() {
        TestifyFeatures.PixelCopyCapture.setEnabled(true)
        rule.setExactness(0.8f)
    }
```

With these settings, I was able to achieve consistently passing tests but not mask layout, copy or size changes (which we want to fail).

It would be my recommendation to try using these values with your tests and see if you have any improvement.

**Comment by matthew-shin-hootsuite** [comment]
@DanielJette thanks for looking into that and I'll give that a try. What does the `TestifyFeatures.PixelCopyCapture.setEnabled(true)` do exactly? Also I also have been seeing some failed cases where I get the following error:

```
androidx.test.espresso.NoActivityResumedException: No activities found. Did you forget to launch the activity by calling getActivity() or startActivitySync or similar?
        at dalvik.system.VMStack.getThreadStackTrace(Native Method)
        at java.lang.Thread.getStackTrace(Thread.java:1720)
        at androidx.test.espresso.base.DefaultFailureHandler.getUserFriendlyError(DefaultFailureHandler.java:12)
        at androidx.test.espresso.base.DefaultFailureHandler.handle(DefaultFailureHandler.java:7)
        at androidx.test.espresso.ViewInteraction.waitForAndHandleInteractionResults(ViewInteraction.java:8)
        at androidx.test.espresso.ViewInteraction.desugaredPerform(ViewInteraction.java:11)
        at androidx.test.espresso.ViewInteraction.perform(ViewInteraction.java:4)
        at androidx.test.espresso.Espresso.closeSoftKeyboard(Espresso.java:1)
        at dev.testify.ScreenshotRule.assertSame(ScreenshotRule.kt:493)
... more stacktrace here related to the test class
```

So for these specific tests we don't seem to run the tests at all, again issue exclusively on M1 devices. 

Lastly I just want to confirm that because the `exactness` uses delta-E algorithm as you mentioned, for example if you set the exactness to 0, when there's changes in layout, copy, or size changes those will still fail, but will ignore any color changes, is that correct?

**Comment by DanielJette** [comment]
@matthew-shin-hootsuite At low exactness values (less than 0.5), the tolerance for differences starts to be so forgiving that even some layout and copy changes will be ignored. At lot of it would depend on the color of the elements. For example, shadows over an already dark background, or dark text over a dark background (`hat` to `hot` on a dark grey screen) might result in a change in pixel from `#5F5F5F` to `#000000` which at a tolerance of 0.5 would pass the check.

In most cases, an `exactness` value in the range of `0.75` through to `0.99` would be safe to use for most layouts. If you use a lot of low-contrast UI (light-on-light or dark-on-dark), then you'll want to use higher values.

**Comment by DanielJette** [comment]
The `NoActivityResumedException` sounds like it could be an issue with the activity under test crashing or being paused while the test is being run. Do you see any other errors or exceptions prior to the test failure?
When you observe the test running, do you see the UI correctly loading?
You can try enabling Layout Inspection Mode which will pause the test immediately prior to taking the screenshot. This will let you visually confirm that the Activity is loading properly and in the desired state.

This shows you how to do that https://ndtp.github.io/android-testify/docs/recipes/layout-inspector

**Comment by matthew-shin-hootsuite** [comment]
@DanielJette so I did some more testing with the recommended pixel copy capture feature enabled, but I actually found that this made the tests fail more than just setting the exactness. I had tests where all the exactness were set at `0.95f` were passing all 6 tests, but then putting in the `@Before` with the pixel copy capture on failed 4 out of the 6 tests that I had. 

So I'm wondering what exactly that does and also doing this brought up another question about the default exactness. In the  `ComposableScreenshotRule` there's a default value of `0.9f` of exactness whereas in the regular `ScreenshotRule` there's no such default value. Is there a reason and will this also affect some of the tests? As the 6 tests I mentioned above in this specific testing are xml views but I do also have most of my tests written in JC

**Comment by matthew-shin-hootsuite** [comment]
> The `NoActivityResumedException` sounds like it could be an issue with the activity under test crashing or being paused while the test is being run. Do you see any other errors or exceptions prior to the test failure? When you observe the test running, do you see the UI correctly loading? You can try enabling Layout Inspection Mode which will pause the test immediately prior to taking the screenshot. This will let you visually confirm that the Activity is loading properly and in the desired state.
> 
> This shows you how to do that https://ndtp.github.io/android-testify/docs/recipes/layout-inspector


yeah I can try those, but again, this issue doesn't seem to be happening on intel computers but just the M1s, so might be worth investigating as well. In the mean time I can see if the mentioned step is something I can try to find a way around it

**Comment by DanielJette** [comment]
@matthew-shin-hootsuite I've been able to successfully get many Testify tests running reliably on M1 hardware and so I believe there is good hope that we can figure out what's going on with your setup!

One thing you can double check is something a colleague of mine just pointed out: the testify documentation has not been updated for M1 support and still mentions using an `x86` emulator image. You should ensure that your emulator image is `arm64` when running on M1 hardware.

**Comment by matthew-shin-hootsuite** [comment]
@DanielJette thanks, yeah so as far as I know, it's not possible to run an `x86` emulator image on an M1 as it will just crash the emulator when you try to launch it so `arm64` one is the only option I have to run (I have tried this). 

I also managed to get all the tests pass by lowering the `exactness` only, but I have to run each test module by module by doing something like `./gradlew :some-module:screenshotTest` and run each module at a time. This is one piece of information I guess I forgot to mention and not sure if it has to do with some of the issues I've been experiencing. Running the entire project at once with `./gradlew screenshotTest` is where I see these inconsistencies I've been talking about where different tests will say the activities has not been started or assertions fail. So I'm wondering if this is an expected behaviour when running the test suite in a multi-module project?

**Comment by DanielJette** [comment]
@matthew-shin-hootsuite 

> So I'm wondering if this is an expected behaviour when running the test suite in a multi-module project?

🤔 That's an interesting piece to the puzzle. I think you might be right. When Testify was first developed, we used it at Shopify on dozens of modules at a time. This was probably around 2015 and we could run `./gradlew screenshotTest` on the root of the project and successfully run the tests on dozens of modules. But, at that time, gradle would run sequentially. Now, I think, gradle will run the commands in parallel. It's entirely likely that the parallel test execution doesn't work correctly and it's mixing up the test contexts. Maybe one test is killing the other activity before it is done running.

A few years ago, Shopify changed up its CI setup so that each module would always be run separately, so if this problem developed, we would not have noticed. Similarly, other projects I've worked on recently we set up to explicitly run the tests separately for each module.

Try running your tests with `./gradlew -Dorg.gradle.parallel=false screenshotTest` as this should force gradle to run the tests one after the other, instead of all-at-once.

**Comment by matthew-shin-hootsuite** [comment]
@DanielJette Thanks a lot for your help. I did run them sequentially (both doing it manually and using the `gradlew` you provided) and they seem to do the trick. For now, I've set all my tests to use `0.8f` exactness and it seems to pass all the tests. Not sure if this is still ideal in the long run, but for now at least it's a workaround for interop between M and Intel chips.

Thanks again!
